doc/NAT/SAI-NAT-API.md

```diff
@@ -0,0 +1,297 @@
+
+nat_attr_count = 5;
+sai_create_nat_fn(
+    &sai_src_nat_obj,
+    switch_id,
+    nat_attr_count,
+    sai_nat_attr_list);
+```
+
+##### Step 4: Create a DNAT object:
+```sh
+sai_nat_attr_list[0].id = SAI_NAT_ATTR_NAT_TYPE;
+sai_nat_attr_list[0].value.u32 = SAI_NAT_TYPE_DESTINATION_NAT;
+
+sai_nat_attr_list[1].id = SAI_NAT_ATTR_FROM_ZONE_ID;
+sai_nat_attr_list[1].value.u8 = 200; 
+
+sai_nat_attr_list[2].id = SAI_NAT_ATTR_TO_ZONE_ID;
+sai_nat_attr_list[2].value.u8 = 100; 
+
+sai_nat_attr_list[3].id = SAI_NAT_ATTR_NAT_ENTRY_LIST;
+sai_nat_attr_list[3].value.objectlist.count = 1; 
+sai_nat_attr_list[3].value.objectlist.list[0] = sai_dst_nat_entry_obj; 
+
+nat_attr_count = 4;
+    sai_create_nat_fn(
+    &sai_dst_nat_obj,
+    switch_id,
+    nat_attr_count,
+    sai_nat_attr_list);
+```
+## NAT Per flow and Zone Counters
+Following per zone statistics is provided by the hardware for NAT feature.
+- DNAT_DISACARDS/SNAT_DISCARDS – If Packet is not TCP/UDP and/or is a fragmentated IP packet. 
+- DNAT_TRANSLATION_NEEDED/SNAT_TRANSLATION_NEEDED – If there is NAT table lookup miss for TCP/UDP packets, then this counter is incremented.
+- DNAT_TRANSLATIONS/SNAT_TRANSLATIONS – If NAT table lookup is a hit, then this counter is incremented.
+
+Counters can be attached to per NAT entry to gather per flow statistics. Per NAT entry stat is enabled in the entry attributes.
+
+> Per zone counters are created using sai_create_nat_counter_fn() API.
+
+NAT counters can be extracted using TAM streaming interface or by using TAM GET API.
+This document proposes to use TAM GET API for time to market reasons. Same NAT counters and data format can be streamed as well using SAI TAM driver.
+
+## GET API for reading NAT Table entries and hit bit
+Besides NAT counters there is a HITBIT table as well which is used to age the NAT entries.
+
+Dynamic NAT entries for a given flow need to be aged once the flow ceases to exist. In this proposal aging is performed by the NOS. NOS retrieves the hitbit table periodically at a configured time aging interface. If the hitbit is not SET for a given entry, the entry is removed from the NAT table using the *sai_remove_nat_entry_fn()* API.
+
+HITBIT table data is retrieved using the TAM proposed generic GET API. 
+Advantage of using TAM GET API is that normalized data for the hitibit data can also be streamed to local NOS, or to an external collector or to multi chip system.
+Some hardware may not perform "clear on read" or NOS may want to control the operation "clear on read". For this reason a separate boolean attribute "CLEAR_ON_READ" is specified 
+
+TAM object is bound to NAT object for stats and hitbit table.
+
+Following is a simple GET API invocation for pulling NAT entry creates under source and destination nat entry objects. One or more objects can be specified in a single invocation. API is name 
+
+
+> sai_tam_telemetry_get_data(switch_id, objlist, clear_on_read, buffer_size, buffer);

```

Comment:
Adding a per entry processing using the flag is too expensive in terms of latency and convergence of NAT. There is already a significant delay in hw update because there is no bulking performed. There is a conscious effort to provide way to bulk updates in SAI API. 
Degenerate case though not suggested can still be achieved by create per entry objects

Reply:
There is already a bulk API get
sai_status_t sai_bulk_get_attribute(
        _In_ sai_object_id_t switch_id,
        _In_ sai_object_type_t object_type,
        _In_ uint32_t object_count,
        _In_ const sai_object_key_t *object_key,
        _Inout_ uint32_t *attr_count,
        _Inout_ sai_attribute_t **attr_list,
        _Inout_ sai_status_t *object_statuses);
which supports optimized get operation
So we only need to add read only activity attribute, and then we can use the bulk get to efficiently read the activity for multiple entries